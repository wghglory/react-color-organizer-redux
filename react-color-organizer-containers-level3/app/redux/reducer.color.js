import C from './constants.color';

export const colorReducer = (state = {}, action) => {
  switch (action.type) {
    case C.ADD_COLOR:
      return {
        id: action.id,
        title: action.title,
        color: action.color,
        timestamp: action.timestamp,
        rating: 0
      };
    case C.RATE_COLOR:
      return (state.id !== action.id) ?
        state :
        // Object.assign({}, state, { rating: action.rating });
        {
          ...state,
          rating: action.rating
        };
    default:
      return state;
  }
};

export const colorsReducer = (state = [], action) => {
  switch (action.type) {
    case C.ADD_COLOR:
      return [
        ...state,
        colorReducer({}, action)
      ];
    case C.RATE_COLOR:
      return state.map(
        c => colorReducer(c, action)
      );
    case C.REMOVE_COLOR:
      return state.filter(
        c => c.id !== action.id
      );
    default:
      return state;
  }
};