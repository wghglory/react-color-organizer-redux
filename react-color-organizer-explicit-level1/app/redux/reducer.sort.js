import C from './constants.color';

export const sortReducer = (state = "SORTED_BY_DATE", action) => {
  switch (action.type) {
    case C.SORT_COLORS:
      return action.sortBy;
    default:
      return state;
  }
};

