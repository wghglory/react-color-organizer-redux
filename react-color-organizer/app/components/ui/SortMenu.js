import React from 'react';
import '../../scss/Menu.scss';
import { PropTypes } from 'prop-types';

const options = {
    date: "SORTED_BY_DATE",
    title: "SORTED_BY_TITLE",
    rating: "SORTED_BY_RATING"
};

/* // pass store via context
const SortMenu = (props, { store }) =>
    (
        <nav className="menu">
            <h1>Sort Colors</h1>
            {Object.keys(options).map((item, i) =>
                <a key={i}
                    href="#"
                    className={(store.getState().sort === options[item]) ? "selected" : null}
                    onClick={e => {
                        e.preventDefault();
                        store.dispatch(sortColors(options[item]));
                    }}>{item}</a>
            )}
        </nav>
    );

SortMenu.contextTypes = {
    store: PropTypes.object
};*/

const SortMenu = ({ sort = "SORTED_BY_DATE", onSelect = f => f }) =>
    (
        <nav className="menu">
            <h1>Sort Colors</h1>
            {Object.keys(options).map((item, i) =>
                <a key={i}
                    href="#"
                    className={(sort === options[item]) ? "selected" : null}
                    onClick={e => {
                        e.preventDefault();
                        onSelect(options[item]);
                    }}>{item}</a>
            )}
        </nav>
    );

SortMenu.propTypes = {
    sort: PropTypes.string,
    onSelect: PropTypes.func
};

export default SortMenu;