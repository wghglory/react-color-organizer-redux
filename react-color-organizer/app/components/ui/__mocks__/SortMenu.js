import React from 'react';
const SortMenuMock = () => <div></div>;
SortMenuMock.displayName = 'SortMenuMock';
export default SortMenuMock;