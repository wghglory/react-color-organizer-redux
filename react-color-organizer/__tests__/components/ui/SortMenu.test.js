import { shallow, mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { compose } from 'redux'
import SortMenu from '../../../app/components/ui/SortMenu'

describe("<Color /> UI Component", () => {

    let wrapper

    it("Renders Correctly", () =>
        compose(expect,toJSON,shallow)(
            <SortMenu sort="SORTED_BY_TITLE" />
        ).toMatchSnapshot()
    )

    it("click does not cause error", () => {
        mount(<SortMenu sort="SORTED_BY_TITLE" />)
            .childAt(2)
            .simulate('click')
    })

    it("onSelect returns correct option", () => {
        const _select = jest.fn()
        const wrapper = mount(<SortMenu onSelect={_select} />)
        wrapper.childAt(3).simulate('click')
        expect(_select.mock.calls[0][0]).toBe("SORTED_BY_RATING")
    })

})
