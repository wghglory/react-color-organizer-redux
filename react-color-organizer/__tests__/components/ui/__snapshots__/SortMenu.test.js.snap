// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color /> UI Component Renders Correctly 1`] = `
<nav
  className="menu"
>
  <h1>
    Sort Colors
  </h1>
  <a
    className={null}
    href="#"
    onClick={[Function]}
  >
    date
  </a>
  <a
    className="selected"
    href="#"
    onClick={[Function]}
  >
    title
  </a>
  <a
    className={null}
    href="#"
    onClick={[Function]}
  >
    rating
  </a>
</nav>
`;
